drivers: modem: add Sierra Wireless HL7800 Modem Driver

rerickson1 opened the pull request:
The driver utilizes the CONFIG_NET_OFFLOAD setting to avoid the
normal handling of IP packets, and instead uses a socket-like
UART interface to handle incoming and outgoing data via AT commands.

jukkar reviewed (changes requested):
Only managed to go through part of the code.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4830 @@
+
+static int hl7800_RX_lock(void)
+{
+	Z_LOG(HL7800_RX_LOCK_LOG_LEVEL, "Locking RX [%p]...", k_current_get());
```
There is something wrong then in the configuration of the log levels etc. No other driver / component in zephyr is using Z_LOG directly as there is no need for that.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4830 @@
+	(void)memset(&sock->dst, 0, sizeof(struct sockaddr));
+}
+
+char *hl7800_sprint_ip_addr(const struct sockaddr *addr)
```
> Tried using `net_sprint_addr()` but it prints the address backwards and causes issues.

The `net_sprint_addr()` expects that the IP address values are in network byte order. If the IP address is printed in reverse order, then it means that the sockaddr values are in host byte order, which would be very surprising.
Both this helper and `net_sprint_addr()` use `net_addr_ntop()` internally, so it sounds a bit weird if they would print the addresses differently.

rerickson1 replied:
I found the difference.  `net_addr_ntop` expects a `struct in_addr *`. My function `hl7800_sprint_ip_addr` expects `struct sockaddr *`.  Leaving my function as a wrapper because it is more convinient to use because it determines the correct `in_addr` or `in6_addr` to use for me.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4830 @@
+{
+	int ret;
+	int numDelims = 7;
+	char *delims[numDelims];
```
I would like that all VLAs (variable length arrays) are removed from this driver.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4841 @@
+#define LOG_DOMAIN modem_hl7800
+#define LOG_LEVEL CONFIG_MODEM_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_DOMAIN);
```
BTW, this can be simplified by just writing
```
LOG_MODULE_REGISTER(modem_hl7800, CONFIG_MODEM_LOG_LEVEL);
```
so normally you do not need to define LOG_DOMAIN or LOG_LEVEL.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4887 @@
+{
+	int ret;
+	int numDelims = 7;
+	char *delims[numDelims];
```
In order to get rid of this VLA, could you have here something like this instead:
```
#define MAX_DELIMS 7
int numDelims = MAX_DELIMS;
char *delims[MAX_DELIMS];
```
The compiler might be able to set this correctly already but just in case better to be in the safe side.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4887 @@
+	}
+
+	/* get new IPv4 addr */
+	int ipv4Len = smStart - addrStart;
```
In zephyr we try to declare variables at the start of the block.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4887 @@
+static bool on_cmd_rtc_query(struct net_buf **buf, uint16_t len)
+{
+	struct net_buf *frag = NULL;
+	size_t str_len = strlen(TIME_STRING_FORMAT);
```
If you use sizeof(), the str_len will be set at compile time.
So `size_t str_len = sizeof(TIME_STRING_FORMAT) - 1;` would be preferred here.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4887 @@
+	/* Ensure the all the expected delimiters are present */
+	size_t offset = TIME_STRING_DIGIT_STRLEN + TIME_STRING_SEPARATOR_STRLEN;
+	size_t i = TIME_STRING_FIRST_SEPARATOR_INDEX;
+	for (; i < TIME_STRING_PLUS_MINUS_INDEX; i += offset) {
```
It is hard to review the kind of code where the variables and code are not separated. It would be helpful for the reviewer if there would be empty line between variables and code. Same applies to other functions in this file.

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4887 @@
+				}
+			}
+
+#ifdef HL7800_LOG_UNHANDLED_RX_MSGS
```
It would be better to use `IS_ENABLED()` macro to make sure that the code inside the the block does not bitrot. This is important to the code blocks that are rarely activated. So here (and similar parts of this file), I would suggest this instead:
```
if (IS_ENABLED(HL7800_LOG_UNHANDLED_RX_MSGS) &&
    !cmd_handled && frag && len > 1) {
...
}
```

jukkar commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4887 @@
+{
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+	if (ictx.uartOn) {
+#ifdef HL7800_IO_LOG
```
I am not very fond of these `#ifdef`'s that are scattered all over the code. You could write these like this instead:
```
#define IO_DBG_LOG(fmt, ...) \
     if (IS_ENABLED(HL7800_IO_LOG) { \
         LOG_DBG(fmt, ##__VA_ARGS__); \
     }
```
and then just do `IO_DBG_LOG("Power OFF the UART");` here. Same for all the extra debug prints you have in this PR.

jukkar reviewed (approved):
LGTM

rerickson1 commented:
> LGTM

Thanks for your patience and thorough review.

carlescufi commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4900 @@
+/* #define HL7800_TX_LOCK_LOG 1 */
+/* #define HL7800_IO_LOG 1 */
+
+#define HL7800_RX_LOCK_DBG_LOG(fmt, ...)                                       \
```
Why do you need those macros? The logging subsystem already has provisions to enable and disable logging of a particular module

rerickson1 replied:
These are extra debug capabilities we want separate from just enabled `DBG`.  These are needed to turn on extra info for digging into interactions in case an advanced bug is found.  Think of this level as `DBG+`.

rlubos commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4900 @@
+static void check_hl7800_awake(void)
+{
+#ifdef CONFIG_MODEM_HL7800_LOW_POWER_MODE
+	bool isCmdRdy = is_cmd_ready();
```
camelCase is not recommended for Zephyr code, snake case should be used instead.

rerickson1 replied:
> camelCase is not recommended for Zephyr code, snake case should be used instead.

Can you point me to this recommendation in any of the zephyr documentation, I couldn't find anything. In the interest of productivity, I plan to leave this as is unless it is deemed critical that I change all of it.

rlubos replied:
From https://www.kernel.org/doc/html/latest/process/coding-style.html, which we follow:
> HOWEVER, while **mixed-case names are frowned upon**, descriptive names for global variables are a must.

As I said, it's not a hard requirement, but a recommendation. Personally, I find it a bit messy, that there are places where even in a single function variables names are following different styles.

jukkar replied:
I missed these camel-case vars in my review, but I agree with @rlubos here, it would be much easier to read/maintain the code if consistent style is followed in the code base. I already gave +1, but @rlubos I suggest that you just give -1 if/when needed or see relevant.

rlubos commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4900 @@
+	int device_services_ind;
+
+	/* modem state */
+	bool allowSleep;
```
Same here, camelCase

rlubos commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4900 @@
+
+	/* modem state */
+	bool allowSleep;
+	bool uartOn;
```
And here

rlubos commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4900 @@
+{
+	int ret;
+	struct net_buf *frag;
+	char dstAddr[sizeof("###.###.###.###")];
```
And here. In general, there are a few more places in this file where camelCase variables mixed in, would be good to sort them out before merge.

rerickson1 committed:
drivers: modem: add Sierra Wireless HL7800 Modem Driver

The driver utilizes the CONFIG_NET_OFFLOAD setting to avoid the
normal handling of IP packets, and instead uses a socket-like
UART interface to handle incoming and outgoing data via AT commands.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rlubos reviewed (approved):
Thank you.

ioannisg commented:
@mike-scott I wonder if you could take a look at this PR that falls under your area of maintenance 👍

mike-scott commented:
> @mike-scott I wonder if you could take a look at this PR that falls under your area of maintenance

Looking today.  Sorry for the delay.

mike-scott reviewed (approved):
I spoke with @rerickson1 on slack and discussed a few things for the future, but this LGTM for merging now.

mike-scott commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4902 @@
+}
+
+#ifdef CONFIG_NEWLIB_LIBC
+int32_t mdm_hl7800_get_local_time(struct tm *tm, int32_t *offset)
```
How is this used?

rerickson1 replied:
We have a public API so the user can get the time from the modem in their app if they wish.

mike-scott reviewed (changes requested):
Woops.  @rerickson1 A quick question on `mdm_hl7800_get_local_time`

mike-scott commented on drivers/modem/hl7800.c:
```diff
@@ -0,0 +1,4902 @@
+}
+#endif
+
+void mdm_hl7800_generate_status_events(void)
```
Also this isn't used as well.

rerickson1 replied:
This is part of a public API in the driver so the app can monitor status.

mike-scott replied:
Is everything in `include/drivers/modem/hl78000.h` part of that public facing API?

mike-scott replied:
ok thanks for explaining
